src/cvrp/ecvrp.py

```diff
@@ -80,58 +91,386 @@ def get_roads(self) -> tuple[tuple[int, ...], ...]:
+
+        self._fitness = self._compute_fitness(self.get_roads())
+        return self._fitness
+
+    def _compute_fitness(self, solution: tuple[tuple[int, ...], ...]) -> float:
+        """
+        Compute the fitness of the solution held in this individual.
+
+        The fitness is, in our case, the maximal time made by a vehicle to
+        delivery the designated clients. Because all the vehicle are leaving
+        the depot at the same time, the fitness will be the vehicle which is
+        taken the most of the time.
+        """
         max_time = 0.
-        for road in self.get_roads():
+        for road in solution:
             current = self._compute_road_fitness(road)
             if current > max_time:
                 max_time = current
-        self._fitness = max_time
-        return self._fitness
+        return max_time
 
-    def _compute_road_fitness(self, road: tuple[int]) -> float:
+    def _compute_road_fitness(self, road: Sequence[int]) -> float:
         road_time = 0.
         latest = road[0]
         for i in road:
             road_time += self.get_instance().get_distance(latest, i)
             latest = i
         return road_time
 
+    def best_place_for(self, solution: list[list[int]], point: int) -> tuple[int, int]:
+        """
+        Find the best place to add a point to delivery.
+
+        This function will be use to find the best place to add a point in a
+        new road at the end of the crossover process.
+        It will return the index of the best place and do not do the insertion
+        """
+        min_fit = float("inf")
+        best_position = (-1, -1)
+        number_roads = len(solution)
+        initial_solution = solution
+
+        for index_road, road in enumerate(initial_solution):
+
+            size_road = len(road)
+            if size_road != 2:
+
+                # we start at index 1 because the index 0 is depot
+                for index_point in range(1, size_road):
+                    road.insert(index_point, point)
+                    fitness = self._compute_fitness(solution)
+                    if fitness < min_fit:
+                        min_fit = fitness
+                        best_position = (index_road, index_point)
+                    road.remove(point)
+            else:
+                road.insert(1, point)
+                fitness = self._compute_fitness(solution)
+                if fitness < min_fit:
+                    min_fit = fitness
+                    best_position = (index_road, 1)
+                road.remove(point)
+
+            if self.__instance.get_ev_count() > number_roads:
+                # If the number of roads is less than the number of vehicules,
+                # the better road can be an additionnal road.
+                depot = self.__instance.get_depot()
+                new_road = [depot, point, depot]
+                solution.append(new_road)
+                fitness = self._compute_fitness(solution)
```

Comment:
Same as before, you give a List where a Tuple is expected

Reply:
done